src/session_impl.cpp

```diff
@@ -2092,29 +2108,40 @@ namespace {
 		, int const flags)
 	{
 		// for now, just pick the first socket with a matching address family
-		// TODO: 3 for proper multi-homed support, we may want to do something
-		// else here. Probably let the caller decide which interface to send over
+		// TODO: remove this function once all callers are updated to specify a socket
 		for (auto& i : m_listen_sockets)
 		{
 			if (!i.udp_sock) continue;
 			if (i.ssl) continue;
 
-			i.udp_sock->send_hostname(hostname, port, p, ec, flags);
-
-			if ((ec == error::would_block
-					|| ec == error::try_again)
-				&& !i.udp_write_blocked)
-			{
-				i.udp_write_blocked = true;
-				ADD_OUTSTANDING_ASYNC("session_impl::on_udp_writeable");
-				i.udp_sock->async_write(std::bind(&session_impl::on_udp_writeable
-					, this, std::weak_ptr<udp_socket>(i.udp_sock), _1));
-			}
+			send_udp_packet_hostname_listen(&i, hostname, port, p, ec, flags);
 			return;
 		}
 		ec = boost::asio::error::operation_not_supported;
 	}
 
+	void session_impl::send_udp_packet_hostname_listen(aux::session_listen_socket* sock
+		, char const* hostname
+		, int const port
+		, span<char const> p
+		, error_code& ec
+		, int const flags)
+	{
+		auto s = static_cast<listen_socket_t*>(sock)->udp_sock;
+
+		s->sock.send_hostname(hostname, port, p, ec, flags);
```

Comment:
I guess each udp socket sets up its own connection to the socks5 proxy, so the source address will match each individual socks5 connection. I suppose it might make sense to talk to the proxy over all interfaces in case we have different bandwidth on different links, but I'm not confident each of these will be able to bind to the same port on the proxy and actually have the interface be transparent. not a big deal for now though.